# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Charge URLs 
file1_url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
file2_url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv"
file3_url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv"

df1 = pd.read_csv(file1_url)
df2 = pd.read_csv(file2_url)
df3 = pd.read_csv(file3_url)

In [6]:
print("DataFrame 1:")
print(df1.columns)

print("DataFrame 2:")
print(df2.columns)

print("DataFrame 3:")
print(df3.columns)

DataFrame 1:
Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='object')
DataFrame 2:
Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Total Claim Amount', 'Policy Type', 'Vehicle Class'],
      dtype='object')
DataFrame 3:
Index(['Customer', 'State', 'Customer Lifetime Value', 'Education', 'Gender',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Total Claim Amount', 'Vehicle Class'],
      dtype='object')


In [8]:
# Renombrar columnas en df3 para que coincidan con df1 y df2
df3.rename(columns={'State': 'ST', 'Gender': 'GENDER'}, inplace=True)

In [9]:
print("Columnas en df1:")
print(df1.columns.tolist())
print("\nColumnas en df2:")
print(df2.columns.tolist())
print("\nColumnas en df3:")
print(df3.columns.tolist())

Columnas en df1:
['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']

Columnas en df2:
['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Total Claim Amount', 'Policy Type', 'Vehicle Class']

Columnas en df3:
['Customer', 'ST', 'Customer Lifetime Value', 'Education', 'GENDER', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Total Claim Amount', 'Vehicle Class']


In [10]:
# Definir el orden estándar de columnas (usamos el orden de df1 como referencia)
orden_comun = [
    'Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
    'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
    'Policy Type', 'Vehicle Class', 'Total Claim Amount'
]

# Reorganizar las columnas de df2 y df3 según el orden común
df2 = df2[orden_comun]
df3 = df3[orden_comun]

In [11]:
print("Columnas en df1:")
print(df1.columns.tolist())
print("\nColumnas en df2:")
print(df2.columns.tolist())
print("\nColumnas en df3:")
print(df3.columns.tolist())

Columnas en df1:
['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']

Columnas en df2:
['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']

Columnas en df3:
['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']


In [12]:
# Concatenar los tres DataFrames
combined_df = pd.concat([df1, df2, df3], ignore_index=True)

In [14]:
combined_df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   9137.000000           9137.000000         9137.000000
mean   37828.820291            110.391266          430.527140
std    30358.716159            581.376032          289.582968
min        0.000000             61.000000            0.099007
25%        0.000000             68.000000          266.996814
50%    34244.000000             83.000000          377.561463
75%    62447.000000            109.000000          546.420009
max    99981.000000          35354.000000         2893.239678

In [15]:
print("Forma del DataFrame combinado:", combined_df.shape)
print("\nPrimeras 5 filas:")
print(combined_df.head())
print("\nColumnas:")
print(combined_df.columns.tolist())

Forma del DataFrame combinado: (12074, 11)

Primeras 5 filas:
  Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Aut

In [16]:
# Convertir nombres de columnas a minúsculas
combined_df.columns = combined_df.columns.str.lower()

# Revisar valores nulos
print("Valores nulos por columna:")
print(combined_df.isnull().sum())

# Revisar tipos de datos
print("\nTipos de datos:")
print(combined_df.dtypes)

Valores nulos por columna:
customer                     2937
st                           2937
gender                       3059
education                    2937
customer lifetime value      2944
income                       2937
monthly premium auto         2937
number of open complaints    2937
policy type                  2937
vehicle class                2937
total claim amount           2937
dtype: int64

Tipos de datos:
customer                      object
st                            object
gender                        object
education                     object
customer lifetime value       object
income                       float64
monthly premium auto         float64
number of open complaints     object
policy type                   object
vehicle class                 object
total claim amount           float64
dtype: object


In [17]:
# Convertir la columna 'customer lifetime value' a tipo numérico
combined_df['customer lifetime value'] = pd.to_numeric(combined_df['customer lifetime value'], errors='coerce')

In [18]:
print("Tipos de datos después de la conversión:")
print(combined_df.dtypes)

Tipos de datos después de la conversión:
customer                      object
st                            object
gender                        object
education                     object
customer lifetime value      float64
income                       float64
monthly premium auto         float64
number of open complaints     object
policy type                   object
vehicle class                 object
total claim amount           float64
dtype: object


In [22]:
# Contar filas donde todas las columnas importantes son nulas
columnas_importantes = [
    'customer', 'st', 'gender', 'education', 'customer lifetime value',
    'income', 'monthly premium auto', 'number of open complaints',
    'policy type', 'vehicle class', 'total claim amount'
]

filas_todas_nulas = combined_df[combined_df[columnas_importantes].isnull().all(axis=1)]
print("Número de filas con TODAS las columnas importantes nulas:", filas_todas_nulas.shape[0])

Número de filas con TODAS las columnas importantes nulas: 2937


In [23]:
# Contar filas que tienen AL MENOS un valor no nulo
filas_con_algun_dato = combined_df[combined_df[columnas_importantes].notnull().any(axis=1)]
print("Número de filas con AL MENOS un valor no nulo:", filas_con_algun_dato.shape[0])

# Contar filas que tienen AL MENOS un valor nulo (pero no todas)
filas_con_datos_parciales = combined_df[
    (combined_df[columnas_importantes].isnull().any(axis=1)) &
    (combined_df[columnas_importantes].notnull().any(axis=1))
]
print("Número de filas con datos parciales (algunos nulos, algunos no):", filas_con_datos_parciales.shape[0])

Número de filas con AL MENOS un valor no nulo: 9137
Número de filas con datos parciales (algunos nulos, algunos no): 2067


In [25]:
print("Tipos de datos:")
print(combined_df.dtypes)

Tipos de datos:
customer                      object
st                            object
gender                        object
education                     object
customer lifetime value      float64
income                       float64
monthly premium auto         float64
number of open complaints     object
policy type                   object
vehicle class                 object
total claim amount           float64
dtype: object


In [28]:
# Copiar el DataFrame original
df_rellenado = combined_df.copy()

# Columnas numéricas (rellenar con mediana)
columnas_numericas = [
    'customer lifetime value', 'income', 'monthly premium auto', 'total claim amount'
]

for col in columnas_numericas:
    mediana = df_rellenado[col].median()
    df_rellenado[col] = df_rellenado[col].fillna(mediana)

# Columnas categóricas (rellenar con 'Desconocido')
columnas_categoricas = [
    'gender', 'education', 'policy type', 'vehicle class'
]

for col in columnas_categoricas:
    df_rellenado[col] = df_rellenado[col].fillna('Desconocido')

In [29]:
print("Valores nulos después del rellenado:")
print(df_rellenado.isnull().sum())

Valores nulos después del rellenado:
customer                     2937
st                           2937
gender                          0
education                       0
customer lifetime value         0
income                          0
monthly premium auto            0
number of open complaints    2937
policy type                     0
vehicle class                   0
total claim amount              0
dtype: int64


In [30]:
df_rellenado.to_csv('combined_clean_data_filled.csv', index=False)

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [31]:
# Cargar el archivo de análisis de marketing
url_marketing = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv"

df_marketing = pd.read_csv(url_marketing)

In [32]:
df_marketing.head()

unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   coverage education effective_to_date employmentstatus gender  ...  \
0     Basic   College        2011-02-18         Employed      M  ...   
1     Basic   College        2011-01-18       Unemployed      F  ...   
2     Basic  Bachelor        2011-02-10         Employed      M  ...   
3  Extended   College        2011-01-11         Employed      M  ...   
4   Premium  Bachelor        2011-01-17    Medical Leave      F  ...   

   number_of_policies     policy_type        policy  renew_offer_type  \
0                   9  Corporate Auto  Corporate L3            Offer3   
1                   1   Personal Auto   Personal L3            Offer4   
2                   2   Personal Auto   Personal L3            Offer3   
3                   2  Corporate Auto  Corporate L3            Offer2   
4                   7   Personal Auto   Personal L2            Offer1   

   sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0          Agent          292.800000  Four-Door Car       Medsize   
1    Call Center          744.924331  Four-Door Car       Medsize   
2    Call Center          480.000000            SUV       Medsize   
3         Branch          484.013411  Four-Door Car       Medsize   
4         Branch          707.925645  Four-Door Car       Medsize   

  vehicle_type month  
0            A     2  
1            A     1  
2            A     2  
3            A     1  
4            A     1  

[5 rows x 27 columns]

In [33]:
print("Forma del DataFrame:", df_marketing.shape)
print("\nColumnas:")
print(df_marketing.columns.tolist())
print("\nPrimeras 5 filas:")
print(df_marketing.head())

Forma del DataFrame: (10910, 27)

Columnas:
['unnamed:_0', 'customer', 'state', 'customer_lifetime_value', 'response', 'coverage', 'education', 'effective_to_date', 'employmentstatus', 'gender', 'income', 'location_code', 'marital_status', 'monthly_premium_auto', 'months_since_last_claim', 'months_since_policy_inception', 'number_of_open_complaints', 'number_of_policies', 'policy_type', 'policy', 'renew_offer_type', 'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size', 'vehicle_type', 'month']

Primeras 5 filas:
   unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   coverage education effective

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [34]:
# Crear la tabla dinámica
pivot_table = df_marketing.pivot_table(
    values='total_claim_amount',  # La columna que queremos sumar
    index='sales_channel',        # La columna que agrupamos (filas)
    aggfunc='sum'                 # La función que aplicamos (sumar)
)

In [35]:
# Redondear a 2 decimales
pivot_table = pivot_table.round(2)

print("Tabla de ingresos totales por canal de ventas:")
print(pivot_table)

Tabla de ingresos totales por canal de ventas:
               total_claim_amount
sales_channel                    
Agent                  1810226.82
Branch                 1301204.00
Call Center             926600.82
Web                     706600.04


La tabla muestra:

Agent: 1,810,226.82
Branch: 1,301,204.00
Call Center: 926,600.82
Web: 706,600.04
Conclusión: Los canales Agent y Branch son los que generaron más ingresos totales, siendo Agent el canal con mayores ventas.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [36]:
# Crear la tabla dinámica
pivot_clv = df_marketing.pivot_table(
    values='customer_lifetime_value',  # La columna que queremos promediar
    index='gender',                    # Agrupar por género en las filas
    columns='education',               # Agrupar por educación en las columnas
    aggfunc='mean'                     # Calcular el promedio
).round(2)  # Redondear a 2 decimales

print("Valor de vida promedio por género y educación:")
print(pivot_clv)

Valor de vida promedio por género y educación:
education  Bachelor  College   Doctor  High School or Below   Master
gender                                                              
F           7874.27  7748.82  7328.51               8675.22  8157.05
M           7703.60  8052.46  7415.33               8149.69  8168.83


Insights (conclusiones):

Nivel educativo más valioso:
Para mujeres (F): High School or Below tiene el valor más alto (8675.22).
Para hombres (M): College tiene el valor más alto (8052.46).
Diferencias por género:
Las mujeres con High School or Below tienen un valor de vida muy superior al resto.
Los hombres con College tienen un valor de vida muy superior al resto.
En general, los hombres tienden a tener valores ligeramente más altos en College y Master.
Nivel educativo menos valioso:
Para ambos géneros, Doctor es el nivel con el valor de vida más bajo (mujeres: 7328.51, hombres: 7415.33).

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [ ]:
# Your code goes here